File: delivery-service/api/src/main/java/org/delivery/api/domain/userOrder/business/UserOrderBusiness.java
package org.delivery.api.domain.userOrder.business;

/*@Business
@RequiredArgsConstructor
@Slf4j
public class UserOrderBusiness {

    private  final UserOrderService userOrderService;
    private final StoreMenuService storeMenuService;
    private final UserOrderConverter userOrderConverter;
    private final UserOrderMenuConverter userOrderMenuConverter;
    private final UserOrderMenuService userOrderMenuService;
    private final StoreService storeService;
    private final StoreMenuConverter storeMenuConverter;
    private final StoreConverter storeConverter;
    private final UserOrderProducer userOrderProducer;
    private final ObjectMapper objectMapper;

    //1. 사용자, 메뉴 id
    //2. userOrder 생성
    //3. userOrderMenu 생성
    //4. 응답 생성
    public UserOrderResponse userOrder(User user, UserOrderRequest body) {

        var storeEntity = storeService.getStoreWithThrow(body.getStoreId());

        var storeMenuEntityList = body.getStoreMenuIdList().stream()
                .map(it -> storeMenuService.getStoreMenuWithThrow(it))
                .collect(Collectors.toList());

        var userOrderEntity = userOrderConverter.toEntity(user, storeEntity, storeMenuEntityList);

        //주문
        var newUserOrderEntity = userOrderService.order(userOrderEntity);

        //맵핑
        var userOrderMenuEntityList = storeMenuEntityList.stream()
                .map(it -> {
                    //menu + user order
                    var userOrderMenuEntity = userOrderMenuConverter.toEntity(newUserOrderEntity, it);
                    return userOrderMenuEntity;
                }).collect(Collectors.toList());

        //주문내역 기록 남기기
        userOrderMenuEntityList.forEach(it -> {
            userOrderMenuService.order(it);
        });
        
        //비동기로 가맹점에 주문알리기
        userOrderProducer.sendOrder(newUserOrderEntity);


        //response
        return userOrderConverter.toResponse(newUserOrderEntity);
    }

    public List<UserOrderDetailResponse> current(User user) {
        var userOrderEntityList = userOrderService.current(user.getId());

        //주문1건씩 처리
        var userOrderDtailResponseList = userOrderEntityList.stream().map(userOrderEntity->{

            //사용자가 주문한 메뉴
            var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList()
                    .stream()
                    .filter(it-> it.getStatus().equals(UserOrderMenuStatus.REGISTERED))
                    .toList();

            var storeMenuEntityList = userOrderMenuEntityList.stream().map(
                    UserOrderMenuEntity::getStoreMenu).collect(Collectors.toList());

            //사용자가 주문한 스토어
            var storeEntity = userOrderEntity.getStore();

            return UserOrderDetailResponse.builder()
                    .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                    .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                    .storeResponse(storeConverter.toResponse(storeEntity))
                    .build();
        }).collect(Collectors.toList());
        return userOrderDtailResponseList;
    }

    public List<UserOrderDetailResponse> history(User user) {
        var userOrderEntityList = userOrderService.history(user.getId());

        //주문1건씩 처리
        var userOrderDtailResponseList = userOrderEntityList.stream().map( userOrderEntity ->{

            *//*
                log.info("사용자의 주문정보 : {}", it);
                try {
                    var jsonString = objectMapper.writeValueAsString(it);
                    log.info("JsonString : {}", jsonString);
                } catch (JsonProcessingException e) {
                    log.info("", e);
                }
            *//*

            //사용자가 주문한 메뉴

            //기존엔 getUserOrderMenu로 쿼리를 보내 데이터를 처리
            //연관관계를 설정했기 때문에 데이터를 간단하게 가져올 수 있음
            //단, 자동으로 데이터를 처리해주는 만큼 잘못처리되는 곳이 있는지 확인필수
            //var userOrderMenuEntityList = userOrderMenuService.getUserOrderMenu(it.getId());
            var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList()
                    .stream()
                    .filter(it -> it.getStatus().equals(UserOrderMenuStatus.REGISTERED))
                    .collect(Collectors.toList());



            var storeMenuEntityList = userOrderMenuEntityList.stream().map(
                    userOrderMenuEntity->{
                        //var storeMenuEntity = storeMenuService.getStoreMenuWithThrow(userOrderMenuEntity.getStoreMenu().getId());
                        //연관관계설정으로 바로 접근가능
                        return userOrderMenuEntity.getStoreMenu();
                    }).collect(Collectors.toList());

            
            //사용자가 주문한 스토어
            var storeEntity = userOrderEntity.getStore();       //연관관계설정으로 바로 접근가능
                    //storeService.getStoreWithThrow(storeMenuEntityList.stream().findFirst().get().getStore().getId());
            return UserOrderDetailResponse.builder()
                    .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                    .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                    .storeResponse(storeConverter.toResponse(storeEntity))
                    .build();
        }).collect(Collectors.toList());
        return userOrderDtailResponseList;

    }

    public UserOrderDetailResponse read(User user, Long orderId) {
        var userOrderEntity = userOrderService.getUserOrderWithOutStatusWithThrow(orderId, user.getId());

        //사용자가 주문한 메뉴
        var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList()
                .stream()
                .filter(it -> it.getStatus().equals(UserOrderMenuStatus.REGISTERED)).collect(Collectors.toList());

        var storeMenuEntityList = userOrderMenuEntityList.stream().map(
                userOrderMenuEntity->{
                    return userOrderMenuEntity.getStoreMenu();
                }).collect(Collectors.toList());

        //사용자가 주문한 스토어
        var storeEntity = userOrderEntity.getStore();

        return UserOrderDetailResponse.builder()
                .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                .storeResponse(storeConverter.toResponse(storeEntity))
                .build();
    }
}*/

File: delivery-service/api/src/main/kotlin/org/delivery/api/domain/userOrder/business/UserOrderBusiness.kt
package org.delivery.api.domain.userOrder.business

import org.delivery.api.common.Log
import org.delivery.api.domain.store.converter.StoreConverter
import org.delivery.api.domain.store.service.StoreService
import org.delivery.api.domain.storeMenu.converter.StoreMenuConverter
import org.delivery.api.domain.storeMenu.service.StoreMenuService
import org.delivery.api.domain.user.model.User
import org.delivery.api.domain.userOrder.controller.model.UserOrderDetailResponse
import org.delivery.api.domain.userOrder.controller.model.UserOrderRequest
import org.delivery.api.domain.userOrder.controller.model.UserOrderResponse
import org.delivery.api.domain.userOrder.converter.UserOrderConverter
import org.delivery.api.domain.userOrder.producer.UserOrderProducer
import org.delivery.api.domain.userOrder.service.UserOrderService
import org.delivery.api.domain.userOrderMenu.converter.UserOrderMenuConverter
import org.delivery.api.domain.userOrderMenu.service.UserOrderMenuService
import org.delivery.common.annotation.Business
import org.delivery.db.userOrderMenu.enums.UserOrderMenuStatus

@Business
class UserOrderBusiness (
    private val userOrderService: UserOrderService,
    private val storeMenuService: StoreMenuService,
    private val userOrderMenuService: UserOrderMenuService,
    private val storeService: StoreService,

    private val userOrderConverter: UserOrderConverter,
    private val userOrderMenuConverter: UserOrderMenuConverter,
    private val storeMenuConverter: StoreMenuConverter,
    private val storeConverter: StoreConverter,

    private val userOrderProducer: UserOrderProducer

){
    companion object: Log

    fun userOrder(
        user: User?,
        body: UserOrderRequest?
    ): UserOrderResponse{
        //가게찾기
        val storeEntity = storeService.getStoreWithThrow(body?.storeId)

        //주문한 메뉴 찾기
        val storeMenuEntityList = body?.storeMenuIdList
            ?.mapNotNull {
                storeMenuService.getStoreMenuWithThrow(it)
            }?.toList()

        val userOrderEntity = userOrderConverter.toEntity(
            user = user,
            storeEntity = storeEntity,
            storeMenuEntityList = storeMenuEntityList
        ).run { userOrderService.order(this) }

        //주문
        val newUserOrderEntity = userOrderService.order(userOrderEntity)

        //맵핑
        val userOrderMenuEntityList = storeMenuEntityList
        ?.map {userOrderMenuConverter.toEntity(
            newUserOrderEntity, it)}
        ?.toList()

        //주문내역 기록남기기
        userOrderMenuEntityList
        ?.forEach {userOrderMenuService.order(it)}

        //비동기로 가맹점에 주문알리기
        userOrderProducer.sendOrder(newUserOrderEntity)

        //response
        return userOrderConverter.toResponse(newUserOrderEntity)
    }

    fun current(
        user: User?
    ): List<UserOrderDetailResponse>?{
        return userOrderService.current(user?.id).map {
            userOrderEntity ->
            //주문1건씩처리
            //사용자가 주문한 메뉴
            val storeMenuEntityList = userOrderEntity.userOrderMenuList?.stream()
                ?.filter { userOrderMenuEntity ->
                    userOrderMenuEntity.status == UserOrderMenuStatus.REGISTERED }
                ?.map { userOrderMenuEntity ->
                    userOrderMenuEntity.storeMenu }
                ?.toList()

           UserOrderDetailResponse(
               userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
               storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
               storeResponse = storeConverter.toResponse(userOrderEntity.store)
           )
         }.toList()
    }

    fun history(
        user: User?
    ): List<UserOrderDetailResponse>?{
        return userOrderService.history(user?.id).map {
                userOrderEntity ->
            //주문1건씩처리
            //사용자가 주문한 메뉴
            val storeMenuEntityList = userOrderEntity.userOrderMenuList?.stream()
                ?.filter { userOrderMenuEntity ->
                    userOrderMenuEntity.status == UserOrderMenuStatus.REGISTERED }
                ?.map { userOrderMenuEntity ->
                    userOrderMenuEntity.storeMenu }
                ?.toList()

            UserOrderDetailResponse(
                userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
                storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
                storeResponse = storeConverter.toResponse(userOrderEntity.store)
            )
        }.toList()
    }

    fun read(
        user: User?,
        orderId: Long?
    ):UserOrderDetailResponse{
        return userOrderService.getUserOrderWithOutStatusWithThrow(orderId, user?.id)
            .let{userOrderEntity ->

            val storeMenuEntityList = userOrderEntity.userOrderMenuList
                ?.stream()
                ?.filter {it.status == UserOrderMenuStatus.REGISTERED }
                ?.map{it.storeMenu }
                ?.toList()
                ?: listOf()

                UserOrderDetailResponse(
                    userOrderResponse = userOrderConverter.toResponse(userOrderEntity),
                    storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
                    storeResponse = storeConverter.toResponse(userOrderEntity.store)
                )
        }
    }
}